x/oracle/abci/proposal.go

```diff
@@ -242,3 +266,109 @@ func (h *ProposalHandler) verifyExchangeRateVotes(
 
 	return nil
 }
+
+func (h *ProposalHandler) generateMedianGasEstimates(
+	ctx sdk.Context,
+	ci cometabci.ExtendedCommitInfo,
+) ([]oracletypes.GasEstimate, error) {
+	gasEstimates := []oracletypes.GasEstimate{}
+
+	for _, vote := range ci.Votes {
+		if vote.BlockIdFlag != cmtproto.BlockIDFlagCommit {
+			continue
+		}
+
+		var valConsAddr sdk.ConsAddress
+		if err := valConsAddr.Unmarshal(vote.Validator.Address); err != nil {
+			h.logger.Error(
+				"failed to unmarshal validator consensus address",
+				"err", err,
+			)
+			return gasEstimates, err
+		}
+		val, err := h.stakingKeeper.GetValidatorByConsAddr(ctx, valConsAddr)
+		if err != nil {
+			h.logger.Error(
+				"failed to get consensus validator from staking keeper",
+				"err", err,
+			)
+			return gasEstimates, err
+		}
+		_, err = sdk.ValAddressFromBech32(val.OperatorAddress)
+		if err != nil {
+			return gasEstimates, err
+		}
+	}
+
+	networks := []string{}
+	// get contracts on registry list
+	params := h.oracleKeeper.GasEstimateKeeper.GetParams(ctx)
+	for _, contract := range params.ContractRegistry {
+		networks = append(networks, contract.Network)
+	}
+
+	for _, network := range networks {
+		networkEstimates := []oracletypes.GasEstimate{}
+
+		for _, vote := range ci.Votes {
+			var voteExt oracletypes.OracleVoteExtension
+			if err := voteExt.Unmarshal(vote.VoteExtension); err != nil {
+				h.logger.Error(
+					"failed to decode vote extension",
+					"err", err,
+				)
+				continue
+			}
+
+			for _, estimate := range voteExt.GasEstimates {
+				if estimate.Network == network {
+					networkEstimates = append(networkEstimates, estimate)
+				}
+			}
+		}
+
+		median := calculateMedian(networkEstimates)
+		gasEstimates = append(gasEstimates, median)
+	}
+
+	return gasEstimates, nil
+}
+
+func (h *ProposalHandler) verifyMedianGasEstimations(
+	injectedEstimates []oracletypes.GasEstimate,
+	generatedEstimates []oracletypes.GasEstimate,
+) error {
+	if len(injectedEstimates) != len(generatedEstimates) {
+		return oracletypes.ErrNonEqualInjVotesLen
+	}
+
+	for i := range injectedEstimates {
+		injectedEstimate := injectedEstimates[i]
+		generatedEstimate := generatedEstimates[i]
+
+		if injectedEstimate.Network != generatedEstimate.Network {
+			return oracletypes.ErrNonEqualInjVotesRates
+		}
+
+		if injectedEstimate.GasEstimation != generatedEstimate.GasEstimation {
+			return oracletypes.ErrNonEqualInjVotesRates
+		}
+	}
+
+	return nil
+}
+
+func calculateMedian(gasEstimates []oracletypes.GasEstimate) (median oracletypes.GasEstimate) {
```

Comment:
Might make sense to move this method into the util file thats already in this directory